### ISSUE TITLE
[Feedback request] Orphan arrangement report

### ISSUE THREAD

**Issue by BuongiornoTexas** [BuongiornoTexas/rsrtools#13]
@JustinAiken

Justin, 

Could I make a low priority request for some feedback on an Orphan arrangement report I've got (I hope) working in 0.3.1. The report checks arrangement ids in a profile against arrangement ids/persistent ids/ids in songs, session mode, and guitar psarc files. It then reports any profile arrangement ids that it can't match against these sources. In effect, it's trying to find profile CDLC data that has become detached from source files.

I'd be interested: to hear what you think from running it against your profiles; if you can think of other sources of arrangement data (persistent ids?) that I may have missed; and in any ideas/suggestions on taking this further. 

The report can be accessed by running songlists (depending on the size of your library it may take a minute or two to scan your dlc), selecting the utility menu, then database reports and choosing option (10). (Apologies for the very ugly formatting). 

I've run this, found about 10 or 12 unattached arrangements and deleted them without impacting either song progress or lesson progress (I have no idea how to assess session or guitar impact, or if there are other places persistent ids might be used). 

Anyhow, it's not urgent, but as and when you have some time, it'd be good to get your thoughts.

Cheers,

Larry

**Comment by JustinAiken** [comment]
I installed the 0.3.1 beta:

```
$ pip3 install -U rsrtools
Collecting rsrtools
  Downloading https://files.pythonhosted.org/packages/26/1b/937dde675a43f422ff5a78929e4fd167cd0e99670d33b51a7faf1c584e65/rsrtools-0.3.1-py3-none-any.whl (119kB)
    100% |████████████████████████████████| 122kB 862kB/s
Requirement already satisfied, skipping upgrade: pydantic>=0.23 in /usr/local/lib/python3.7/site-packages (from rsrtools) (0.24)
Requirement already satisfied, skipping upgrade: pycryptodome>=3.7.3 in /usr/local/lib/python3.7/site-packages (from rsrtools) (3.8.1)
Requirement already satisfied, skipping upgrade: simplejson>=3.16 in /usr/local/lib/python3.7/site-packages (from rsrtools) (3.16.0)
Requirement already satisfied, skipping upgrade: toml>=0.10.0 in /usr/local/lib/python3.7/site-packages (from rsrtools) (0.10.0)
Installing collected packages: rsrtools
  Found existing installation: rsrtools 0.2.2
    Uninstalling rsrtools-0.2.2:
      Successfully uninstalled rsrtools-0.2.2
Successfully installed rsrtools-0.3.1
```

Went to my working directory and tried to run songlists:

```ruby
$ songlists .

No arrangement data in database. Running full scan to load data.
Arrangement 10, The Cribs, We Share the Same Skies
.... <more stuff> ...
Arrangement 840, Steel Panther, Eyes of a Panther
Traceback (most recent call last):
  File "/usr/local/bin/songlists", line 10, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.7/site-packages/rsrtools/songlists/songlists.py", line 887, in main
    main.song_list_cli()
  File "/usr/local/lib/python3.7/site-packages/rsrtools/songlists/songlists.py", line 572, in song_list_cli
    self._cli_full_scan()
  File "/usr/local/lib/python3.7/site-packages/rsrtools/songlists/songlists.py", line 559, in _cli_full_scan
    mtime = self._arr_db.scan_arrangements(last_modified=None, show_progress=True)
  File "/usr/local/lib/python3.7/site-packages/rsrtools/songlists/database.py", line 809, in scan_arrangements
    for sql_values in Scanner().db_entries(last_modified=last_modified):
  File "/usr/local/lib/python3.7/site-packages/rsrtools/songlists/scanner.py", line 497, in db_entries
    for arrangement in self._arrangement_rows(scan_path, False):
  File "/usr/local/lib/python3.7/site-packages/rsrtools/songlists/scanner.py", line 446, in _arrangement_rows
    psarc.arc_data(index), expect_name, last_modified
  File "/usr/local/lib/python3.7/site-packages/rsrtools/songlists/scanner.py", line 378, in _arrangement_data
    arrangement[ListField.TUNING.value] = self._get_tuning(sub_dict)
  File "/usr/local/lib/python3.7/site-packages/rsrtools/songlists/scanner.py", line 203, in _get_tuning
    raise ValueError(f"Invalid or unknown tuning: {tuning}.")
ValueError: Invalid or unknown tuning: -8-6-6-6-8-8.
```

Looks like it doesn't like very low tunings... Or might be a cdlc bass track using guitar as 5 string bass? Not sure which song it's crashing on..

**Comment by JustinAiken** [comment]
Ah added some prints to debug - it's the `staisofa_rhythm2` for Staind - So far Away. 

Relevant bit:

```json
        "Tuning": {
          "string0": -8,
          "string1": -6,
          "string2": -6,
          "string3": -6,
          "string4": -8,
          "string5": -8
        },
```

Which is an odd [Ab,Db,Ab,Db,Eb,Ab tuning](https://theriffrepeater.com/rocksmith-2014-dlc-962016-staind-its-been-awhile-so-far-away-outside/)

Could maybe change https://github.com/BuongiornoTexas/rsrtools/blob/a7c81a5376bf6ebfeda311b899fd64d56c98d076/rsrtools/songlists/scanner.py#L200-L203 to just `return "Oddball Tuning"` for an `else` instead of the `raise`

**Comment by BuongiornoTexas** [comment]
Blast! - I thought I had implemented exactly that after adding all of the CFSM tunings. I'll get that sorted and push a new release later today or tomorrow. I might use "Custom Tuning" to be consistent with RS.

Thanks for picking it up.

**Comment by BuongiornoTexas** [comment]
I had time over a coffee - 0.3.2 reports Unknown Custom Tuning for this case now.

**Comment by BuongiornoTexas** [comment]
@JustinAiken Sorry - I just realised you may not be following this actively (I'm come from the paper, pen  and phone call generation, so I'm still getting used to the "ats"). 

Anyhoo, I'm hoping 0.3.2. has addressed the issue. No hurry on feedback - I'm happy not doing too much development at the moment - new drop pedal yesterday, so I'm having fun experimenting and enjoying being able to play low tunings without spaghetti strings!

**Comment by JustinAiken** [comment]
Sorry for the delay! I get @s (github's my homepage), I've just been slowing compiling feedback.  There's a bit list of arrangement ids the report gives, and I'm trying to track down which ones are true orphans, and which ones are for ones I'd temporarily removed from the library.  

- No more tuning errors...
- I do see `ValueError: Arrangement FullName in JSON does not match JSON arrangement name in file` on some CDLC, which is probably the bad CDLC's fault, but it blocks the process instead of skipping over

**Comment by BuongiornoTexas** [comment]
Hmmm - I can do a couple of things with this.

1) Make the error message a bit more informative - I hadn't been expecting it to trigger at all, but you are probably right about it being a CDLC issue, and it may also be a case matching issue (which will bring it's own complications!). 

I'll add in details to the error dialogue to allow trapping of the specific file/arrangement.

2) As you suggest, I can trap and exclude the arrangement from the database - though I'm not sure if this is the right thing to do or not? (If the CDLC works, is it a problem if it doesn't follow the Ubisoft standard?) - probably worth holding off on this until we understand the underlying cause?

Also, if you have the CDLCs that you have temporarily removed from Rockmsith parked in a single directory tree, I think it I could put together a quick and dirty script to dump all of the arrangement ids pretty quickly.

**Comment by BuongiornoTexas** [comment]
I've done a quick and dirty update to the exception. If you can do a github install, we can get some more information on the problem before I do a full fix and rollout to pypi:

``pip install git+https://github.com/BuongiornoTexas/rsrtools --upgrade``

This should tell us if it is a case folding problem or something else, and give me a better handle on implementing a fix.

**Comment by BuongiornoTexas** [comment]
Hi @JustinAiken, I'm not sure if you have played around with this any further - the fix is now in the 0.3.4 release, and gives information on the odd tunings so that they can be added to the database in due course.

**Comment by BuongiornoTexas** [comment]
Closed due to lack of activity.
